More sensible default language pair

MaathavanJkr opened the pull request:
more sensible default language pair when apertium.org is loaded. I think it might work!

MaathavanJkr commented:
@sushain97 how can i run this code?

sushain97 commented:
What error did you run into when following the instructions in the README?

MaathavanJkr commented:
I ran it. :), But

sushain97 commented:
But...?

MaathavanJkr commented:
Is it should select the browser preferred language??. and can it be a seperate js file?

MaathavanJkr commented:
okay im done i think. :)

MaathavanJkr commented:
Now the default translator pair is first available user's browser preference language pair. :) or it will select english as default. :)

MaathavanJkr commented:
https://maathavanjkr.github.io/ here you can see it in action. :)

MaathavanJkr commented:
#343

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,23 @@ function getPairs() /*: JQueryPromise<any> */ {
-            break;
+        // Default for new users is first available Browser Preference Language pair
+        var prefLang;
+        if(languages[navigator.languages[0]] !== undefined) {
```
What is the availability of the `navigator.languages` API? html-tools currently maintains compatibility with a relatively old set of browsers.

Secondly, why is this not possible with a loop...?

MaathavanJkr replied:
It is available for all Chrome, Safari and Opera Versons and Firefox 1.0

MaathavanJkr replied:
Yeahh Its possible. :).

MaathavanJkr replied:
Loop.. Done. :)

sushain97 replied:
What about IE though...?

MaathavanJkr replied:
IE??

MaathavanJkr replied:
Internet Explorer, 11.0

MaathavanJkr replied:
Okay.. Now it will support IE also. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,23 @@ function getPairs() /*: JQueryPromise<any> */ {
+            prefLang = navigator.languages[2];
         }
+        else {
+            prefLang = 'en';
```
The preference should be the `locale` when available.

MaathavanJkr replied:
is it possible that locale will available for all time?

MaathavanJkr replied:
done :)..

MaathavanJkr commented:
> This appears to change to the browser-preferred language _even if it isn't an available for translation_. This is incorrect behaviour.

can u please give an example language for this?

MaathavanJkr commented:
> This appears to change to the browser-preferred language _even if it isn't an available for translation_. This is incorrect behaviour.

Its Done I think. :) anymore changes? :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,33 @@ function getPairs() /*: JQueryPromise<any> */ {
-            curDstLang = pairs[srcLang][0];
-            break;
+        // Default for new users is first available Browser Preference Language pair
+        var brwsLangs = navigator.languages; // Chrome, Mozilla and Safari
```
```suggestion
        var browserLangs = navigator.languages; // Chrome, Mozilla and Safari
```

MaathavanJkr replied:
Done. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,33 @@ function getPairs() /*: JQueryPromise<any> */ {
+        // Default for new users is first available Browser Preference Language pair
+        var brwsLangs = navigator.languages; // Chrome, Mozilla and Safari
+        var prefLang;
+        $.each(brwsLangs, function(index) {
```
`$.each` also gives you the actual element, using that should work fine here.

MaathavanJkr replied:
already $.each is used. I dont understand?

sushain97 replied:
The callback receives each value, there's no need to re-index the values from the array. Please look at the documentation.

MaathavanJkr replied:
Aah okay okay. :)

MaathavanJkr replied:
Gotcha! :)

MaathavanJkr replied:
Done. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,33 @@ function getPairs() /*: JQueryPromise<any> */ {
+            }
+        });
+        if(typeof prefLang === 'undefined') {
+            if(!navigator.userlanguage && !navigator.browserlanguage) {
```
Please switch these branches for increased readability.

MaathavanJkr replied:
Increased readability?

sushain97 replied:
Yes, the code will be easier to understand if you switch these branches.

MaathavanJkr replied:
switch these branches. How? I dont understand.

MaathavanJkr replied:
okay got it!:)

MaathavanJkr replied:
Done. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,33 @@ function getPairs() /*: JQueryPromise<any> */ {
+                curSrcLang = locale;
+            }
+            else {
+                var bLang = navigator.userlanguage || navigator.browserlanguage; // Internet Explorer
```
```suggestion
                var ieLang = navigator.userlanguage || navigator.browserlanguage; // Internet Explorer
```

MaathavanJkr replied:
Done. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,33 @@ function getPairs() /*: JQueryPromise<any> */ {
+            }
+            else {
+                var bLang = navigator.userlanguage || navigator.browserlanguage; // Internet Explorer
+                curSrcLang = iso639CodesInverse[bLang];
```
What happens when `bLang` isn't in the available source languages?

MaathavanJkr replied:
oops, i forgot it. :(

MaathavanJkr replied:
Done. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,33 @@ function getPairs() /*: JQueryPromise<any> */ {
+        var brwsLangs = navigator.languages; // Chrome, Mozilla and Safari
+        var prefLang;
+        $.each(brwsLangs, function(index) {
+            if(languages[brwsLangs[index].substr(0, 2)] !== undefined) {
```
Why are you always taking the substring? There are some languages that don't have 2-letter codes.

MaathavanJkr replied:
https://www.metamodpro.com/browser-language-codes in this list i didnt find any languages with 3 letter codes. I used Because when en-US is preferred language, there wont be a language.so i trimmed en

MaathavanJkr replied:
When en-US is browser prefered language. it doesnt recognize it and goes to the next lang thats why

sushain97 replied:
I don't consider that list particularly canonical... Can you do something less brittle like splitting on `-`? I'd save that into a variable so you don't have to split twice.

MaathavanJkr replied:
Okay. :) thats fine. I will do it. :)

MaathavanJkr replied:
Done. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,33 @@ function getPairs() /*: JQueryPromise<any> */ {
+        var prefLang;
+        $.each(brwsLangs, function(index) {
+            if(languages[brwsLangs[index].substr(0, 2)] !== undefined) {
+                curSrcLang = iso639CodesInverse[brwsLangs[index].substr(0, 2)];
```
What if none of the languages are valid source languages? This approach seems like it would leave the application in some crazy state.

MaathavanJkr replied:
Oops I for got it here. :)

MaathavanJkr replied:
it will move to next step below, set to Locale. :).

MaathavanJkr replied:
Nope, See below it will set to locale.

MaathavanJkr replied:
Done. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,33 @@ function getPairs() /*: JQueryPromise<any> */ {
+        });
+        if(typeof prefLang === 'undefined') {
+            if(!navigator.userlanguage && !navigator.browserlanguage) {
+                curSrcLang = locale;
```
Are you certain that `locale` has been set at this point? I'm not sure it's guaranteed.

MaathavanJkr replied:
not guaranteed? is it will not available some times?

MaathavanJkr replied:
added eng when locale also unavailable. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1242,6 +1265,43 @@ function autoSelectDstLang() {
+function checkLangPairAvailable(lang) {
+    if(languages[lang] !== undefined) {
+        curSrcLang = iso639CodesInverse[lang];
+        autoSelectDstLang();
```
A function called `checkLangPairAvailable` should not have any side effects, i.e. it shouldn't be flapping around the state of the UI by changing state.

MaathavanJkr replied:
Will it create side effects?

MaathavanJkr replied:
Okay working on it.:)

MaathavanJkr replied:
Done. I think! :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1242,6 +1265,43 @@ function autoSelectDstLang() {
+    if(languages[lang] !== undefined) {
+        curSrcLang = iso639CodesInverse[lang];
+        autoSelectDstLang();
+        if(curDstLang !== undefined) {
```
This is all equivalent to `return curDstLang !== undefined`

MaathavanJkr replied:
Aahh, yup thats right. :)

MaathavanJkr replied:
Done. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1242,6 +1265,43 @@ function autoSelectDstLang() {
+// Get the language code if - exist
+function getLangCode(lang) {
+    if(lang.includes('-')) {
+        return lang.substr(0, lang.indexOf('-'));
```
`.split('-', 2)` works perfectly fine.

MaathavanJkr replied:
Okay.

MaathavanJkr replied:
Done. :)

MaathavanJkr commented:
@sushain97 anymore changes? :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,35 @@ function getPairs() /*: JQueryPromise<any> */ {
-            curSrcLang = srcLang;
-            curDstLang = pairs[srcLang][0];
-            break;
+        // Default for new users is first available Browser Preference Language pair
```
Could you move all of this logic into a separate function?

MaathavanJkr replied:
Done. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1242,6 +1265,36 @@ function autoSelectDstLang() {
     $('#dstLangSelect').val(curDstLang).change();
 }
 
+// Check if the Language pair Exist or Not
```
These three functions can go inside the separate function you create.

MaathavanJkr replied:
Done :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,35 @@ function getPairs() /*: JQueryPromise<any> */ {
+        // Default for new users is first available Browser Preference Language pair
+        var browserLangs = navigator.languages; // Chrome, Mozilla and Safari
+        var prefLang;
+        $.each(browserLangs, function(i, langVal) {
```
```suggestion
        $.each(browserLangs, function(i, lang) {
```

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,35 @@ function getPairs() /*: JQueryPromise<any> */ {
-            break;
+        // Default for new users is first available Browser Preference Language pair
+        var browserLangs = navigator.languages; // Chrome, Mozilla and Safari
+        var prefLang;
```
```suggestion
        var prefSrcLang;
```

MaathavanJkr replied:
Done. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,35 @@ function getPairs() /*: JQueryPromise<any> */ {
+                return false;
+            }
+        });
+        if(typeof prefLang === 'undefined') {
```
```suggestion
        if(prefSrcLang) {
```

MaathavanJkr replied:
Done :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,35 @@ function getPairs() /*: JQueryPromise<any> */ {
+            }
+        });
+        if(typeof prefLang === 'undefined') {
+            if(navigator.userlanguage !== undefined || navigator.browserlanguage !== undefined) {
```
`navigator.userlanguage || navigator.browserlanguage` can be saved to a variable and then used here and below.

MaathavanJkr replied:
Done :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,35 @@ function getPairs() /*: JQueryPromise<any> */ {
+                    curSrcLang = iso639CodesInverse[ieLang];
+                }
+                else {
+                    noPrefSrcLang();
```
Instead of doing this twice, at the end, check `if(!prefSrcLang)` and do it once. Then, you can eliminate the extra `noPrefSrcLang` function as well.

MaathavanJkr replied:
Done :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1242,6 +1265,36 @@ function autoSelectDstLang() {
+
+// Get the language code if - exist
+function getLangCode(lang) {
+    return lang.split('-', 2)[0];
```
Apertium actually supports language variants so we shouldn't delete the second part. Instead we should convert e.g. `en-US` to `en_US`. The latter is what Apertium uses.

MaathavanJkr replied:
done. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,35 @@ function getPairs() /*: JQueryPromise<any> */ {
+            if(navigator.userlanguage !== undefined || navigator.browserlanguage !== undefined) {
+                var ieLang = getLangCode(navigator.userlanguage || navigator.browserlanguage); // Internet Explorer
+                if(checkLangPairAvailable(ieLang)) {
+                    curSrcLang = iso639CodesInverse[ieLang];
```
Instead, set `prefSrcLang`.

MaathavanJkr replied:
Done :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +447,35 @@ function getPairs() /*: JQueryPromise<any> */ {
+            }
+        }
+        else {
+            curSrcLang = iso639CodesInverse[prefLang];
```
Instead, set `prefSrcLang`. Then, set `curSrcLang` once at the end if you have `prefSrcLang`.

MaathavanJkr replied:
Done :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1242,6 +1265,36 @@ function autoSelectDstLang() {
 
+// Check if the Language pair Exist or Not
+function checkLangPairAvailable(lang) {
+    if(languages[lang] !== undefined) {
```
```suggestion
    if(languages[lang]) {
```

MaathavanJkr replied:
Done :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1242,6 +1265,36 @@ function autoSelectDstLang() {
+function checkLangPairAvailable(lang) {
+    if(languages[lang] !== undefined) {
+        var checkDstLang = pairs[iso639CodesInverse[lang]][0];
+        return checkDstLang !== undefined;
```
Actually, I think this whole function can be `return languages[lang] && pairs[iso639CodesInverse[lang]].length`.

MaathavanJkr replied:
Yeah thats short. :)

MaathavanJkr replied:
Done :)

MaathavanJkr replied:
and just return languages[lang] && pairs[iso639CodesInverse[lang]] is enough i think

MaathavanJkr commented:
What can i do with the Checks?

sushain97 commented:
Have you clicked on the links? They’ll tell you what the errors are.

On Sat, Dec 14, 2019 at 8:44 PM Maathavan <notifications@github.com> wrote:

> What can i do with the Checks?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/apertium/apertium-html-tools/pull/352?email_source=notifications&email_token=ABAEI7QL22URDVQRM2LN44DQYWY3HA5CNFSM4JY4WZSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEG4RKNI#issuecomment-565777717>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAEI7UMZXOFZ7GH6MS2GMTQYWY3HANCNFSM4JY4WZSA>
> .
>
-- 

[image: Sushain Cherivirala]
*Sushain K. Cherivirala*
Stripe, Software Engineer
Stanford University, M.S. in Computer Science '19
Carnegie Mellon University, B.S. in Computer Science '18
(713) 992-4043 | www.skc.name

MaathavanJkr commented:
Yeahh its telling that, some variables doesnt exist, which are included in other script files

sushain97 commented:
The import and export declarations at the top/bottom of the files should
help with that.

On Sat, Dec 14, 2019 at 8:48 PM Maathavan <notifications@github.com> wrote:

> Yeahh its telling that, some variables doesnt exist, which are included in
> other script files
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/apertium/apertium-html-tools/pull/352?email_source=notifications&email_token=ABAEI7SAS3XZ4UZNAVTKE7TQYWZKHA5CNFSM4JY4WZSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEG4RL4I#issuecomment-565777905>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAEI7SLE4G3FDXEPJGFGILQYWZKHANCNFSM4JY4WZSA>
> .
>
-- 

[image: Sushain Cherivirala]
*Sushain K. Cherivirala*
Stripe, Software Engineer
Stanford University, M.S. in Computer Science '19
Carnegie Mellon University, B.S. in Computer Science '18
(713) 992-4043 | www.skc.name

MaathavanJkr commented:
Okay! :)

MaathavanJkr commented:
I dont understand why these checks are failing now. :(

sushain97 commented:
I’m not in a position to look closely right now but I suggest googling the
names of the rules as well as setting up eslint to work locally.

On Sat, Dec 14, 2019 at 10:13 PM Maathavan <notifications@github.com> wrote:

> I dont understand why these checks are failing now. :(
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/apertium/apertium-html-tools/pull/352?email_source=notifications&email_token=ABAEI7TDKZ65DOOV45HMPGTQYXDKDA5CNFSM4JY4WZSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEG4SIYY#issuecomment-565781603>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAEI7STFLWCQK3A3XUREZTQYXDKDANCNFSM4JY4WZSA>
> .
>
-- 

[image: Sushain Cherivirala]
*Sushain K. Cherivirala*
Stripe, Software Engineer
Stanford University, M.S. in Computer Science '19
Carnegie Mellon University, B.S. in Computer Science '18
(713) 992-4043 | www.skc.name

MaathavanJkr commented:
The errors are like Property `userlanguage` is missing in `Navigator` [1]. Property `replace` is missing in `Lang`.

sushain97 commented:
The former is probably because the Flow types don’t include IE stuff.
You’ll probably want to cast to any to silence them. The latter I’m not
sure about without more context.

On Sat, Dec 14, 2019 at 10:17 PM Maathavan <notifications@github.com> wrote:

> The errors are like Property userlanguage is missing in Navigator [1].
> Property replace is missing in Lang.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/apertium/apertium-html-tools/pull/352?email_source=notifications&email_token=ABAEI7SIPPNVQ37OTUKSVVTQYXDYBA5CNFSM4JY4WZSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEG4SKKQ#issuecomment-565781802>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAEI7RROYRWG3BPEIH4BWDQYXDYBANCNFSM4JY4WZSA>
> .
>
-- 

[image: Sushain Cherivirala]
*Sushain K. Cherivirala*
Stripe, Software Engineer
Stanford University, M.S. in Computer Science '19
Carnegie Mellon University, B.S. in Computer Science '18
(713) 992-4043 | www.skc.name

MaathavanJkr commented:
The Docker Cloud link isnt working.

sushain97 commented:
I don't know why the links aren't working but I took a quick look at it and it seems like something broken on Dockercloud's end...

sushain97 commented on assets/js/translator.js:
```diff
@@ -1,5 +1,6 @@
 // @flow
-
+var languages;
+var iso639CodesInverse;
```
These should be `/*:: import`d from the correct modules, not re-declared like this,

sushain97 commented on assets/js/translator.js:
```diff
@@ -447,12 +448,8 @@ function getPairs() /*: JQueryPromise<any> */ {
-            curDstLang = pairs[srcLang][0];
-            break;
-        }
+        // Setting Default language for new user
```
```suggestion
```

Let's try to make sure our comments add value!

sushain97 commented on assets/js/translator.js:
```diff
@@ -1252,6 +1249,52 @@ function autoSelectDstLang() {
+    // Default for new users is first available Browser Preference Language pair
+    var browserLangs = window.navigator.languages; // Chrome, Mozilla and Safari
+    var ieLang = window.navigator.userlanguage || window.navigator.browserlanguage; // Internet Explorer
+    var prefSrcLang;
```
I don't think we're preserving the behavior anymore where a user who comes back to the site gets the same source language they switched to. Are we?

MaathavanJkr replied:
Yeahh... i didnt notice it. I can Fix it. :)

MaathavanJkr replied:
Done. :)

sushain97 replied:
I don't see how? `curSrcLang = iso639CodesInverse[prefSrcLang];` will happen with `prefSrcLang` not set.

MaathavanJkr replied:
Where?

MaathavanJkr replied:
'if(!prefSrcLang) {
        ....
    }
    else {
        curSrcLang = iso639CodesInverse[prefSrcLang];
        curDstLang = pairs[curSrcLang][0];
    }'

MaathavanJkr replied:
its happen after prefSrcLang is set

sushain97 replied:
ah, right

sushain97 commented on assets/js/translator.js:
```diff
@@ -1252,6 +1249,52 @@ function autoSelectDstLang() {
+    var browserLangs = window.navigator.languages; // Chrome, Mozilla and Safari
+    var ieLang = window.navigator.userlanguage || window.navigator.browserlanguage; // Internet Explorer
+    var prefSrcLang;
+    var i;
```
This var can be inlined below.

MaathavanJkr replied:
inlined below?

sushain97 replied:
it doesn't need to be declared separately, it can be inlined inside the loop

MaathavanJkr replied:
Okay. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1252,6 +1249,52 @@ function autoSelectDstLang() {
+    }
+
+    // Set curSrcLang as locale or first available pair, if locale not available
+    function noPrefSrcLang() {
```
Please inline this function. It's only used once and is relatively small.

MaathavanJkr replied:
Yeahh its it should be. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1252,6 +1249,52 @@ function autoSelectDstLang() {
+            break;
+        }
+    }
+    // Get the language code if - exist
```
This function converts the language code from browser -> apertium format.

MaathavanJkr replied:
Yup. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1252,6 +1249,52 @@ function autoSelectDstLang() {
+    // Default for new users is first available Browser Preference Language pair
+    var browserLangs = window.navigator.languages; // Chrome, Mozilla and Safari
+    var ieLang = window.navigator.userlanguage || window.navigator.browserlanguage; // Internet Explorer
+    var prefSrcLang;
```
```suggestion
    var prefSrcLang;

```

sushain97 commented on assets/js/translator.js:
```diff
@@ -1252,6 +1249,52 @@ function autoSelectDstLang() {
+            prefSrcLang = browserLang;
+            break;
+        }
+    }
```
```suggestion
    }

```

sushain97 commented on assets/js/translator.js:
```diff
@@ -1252,6 +1249,52 @@ function autoSelectDstLang() {
+                prefSrcLang = iePrefLang;
+            }
+        }
+    }
```
```suggestion
    }

```

sushain97 commented on assets/js/translator.js:
```diff
@@ -1252,6 +1249,52 @@ function autoSelectDstLang() {
+    }
+    if(!prefSrcLang) {
+        noPrefSrcLang();
+    }
```
```suggestion
    }

```

MaathavanJkr commented:
I think imported the Variables correctly, But the eslint still returns error.:(

sushain97 commented on assets/js/translator.js:
```diff
@@ -1281,5 +1323,6 @@ function setRecentDstLangs(langs /*: string[] */) {
 /*:: import {ENTER_KEY_CODE, HTTP_BAD_REQUEST_CODE, HTTP_OK_CODE, SPACE_KEY_CODE, XHR_DONE, XHR_LOADING} from "./util.js" */
 /*:: import {persistChoices, restoreChoices} from "./persistence.js" */
 /*:: import {localizeInterface, getLangByCode, getDynamicLocalization, locale, iso639Codes, langDirection} from "./localization.js" */
+/*:: import {languages, iso639CodesInverse} from "./localization.js" */
```
Please combine this with the above. Also, the problem might be that that `localization.js` doesn't export them. Have you verified they do?

MaathavanJkr replied:
When i combined them eslint check said to too many in one line error.

MaathavanJkr replied:
Oops. Oh I have to export them. :)

MaathavanJkr replied:
What to do combine or not?

sushain97 replied:
You can split it into multiple lines without having multiple `import` statements. There's an example of that on L1317.

MaathavanJkr replied:
i correctly exported and imported i think, but its still returns error

sushain97 replied:
Ah, right, I think it also needs to be added to the `global` declaration at the top to satisfy ESLint. The exports/imports are for Flow.

MaathavanJkr replied:
Like i did before?

sushain97 replied:
Not exactly. I'm referring to https://github.com/apertium/apertium-html-tools/pull/352/commits/27ecdc01b1c5835b2d4ca130afee5b646e288cfe#diff-8331c23a07e5eb815c5b4393e81715efL27.

MaathavanJkr replied:
Okay got it. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1272,6 +1267,52 @@ function setRecentDstLangs(langs /*: string[] */) {
+    }
+
+    if(!prefSrcLang) {
+        if(ieLang) {
```
These conditionals can be combined.

MaathavanJkr replied:
Done. :)

MaathavanJkr commented:
@sushain97  I cant still get Dockercloud error. :(

sushain97 commented:
You can ignore the Dockercloud error.

I want to make some minor tweaks before I merge this but the code looks like it should do the right thing now.

@jonorthwash do you think you could test that it

a) defaults to using the user's browser language on their first visit
b) preserves your selection if you switch languages and revisit the site (not refreshing but revisiting entirely)

(b) was broken in the initial version and (a) didn't work on IE.

MaathavanJkr commented:
> (b) was broken in the initial version and (a) didn't work on IE.
broken?

MaathavanJkr commented:
https://maathavanjkr.github.io/ can check it here. i just updated it :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1307,7 +1307,7 @@ function setDefaultSrcLang() {
     // Convert language code from browser -> apertium format.
     function changeLangCode(lang) {
-        return lang.replace('-', '_');
+        return lang.split('-')[0];
```
Why did this get switched back again? I thought we discussed that we need to preserve the variant since Apertium supports them.

MaathavanJkr replied:
Replied below.

MaathavanJkr commented:
I cant find en_US like codes in languages or ISO639codes, therefore it doesnt worked on IE, when i checked So i changed to 'lang.split("-")[0]', is ther anything needed to be changed?

sushain97 commented:
> I cant find en_US like codes in languages or ISO639codes, therefore it doesnt worked on IE, when i checked So i changed to 'lang.split("-")[0]', is ther anything needed to be changed?

https://tools.ietf.org/html/bcp47 page 80 says that the variants will have dashes. We shouldn't discard them by doing the split.

Are you saying that IE uses only `en-US` and not `en`? In that case, we should do a special case for IE where we test both (with variant and without).

BTW, I pushed some tweaks to this branch.

sushain97 commented:
Ah, I see what you're saying. `pairs[iso639CodesInverse[lang]]` isn't the right way to do this because it doesn't handle the variant.

MaathavanJkr commented:
Nope IE uses the Windows set language. In my computer its set as en-US. thats why i got error. And i just tested it with Google chrome by setting en-US, Its also didnt work. Then i saw there isnt any defined as en_US.

MaathavanJkr commented:
> Ah, I see what you're saying. `pairs[iso639CodesInverse[lang]]` isn't the right way to do this because it doesn't handle the variant.

Yup. :)

sushain97 commented:
OK, I fixed the conversion. I'll leave it to you to clean up the lint errors ;)

Now, what you need to do is say that if `aaa_bbb` is the only thing IE provides, we check `aaa_bbb` first and then `aaa`. My Chrome provides both `en-US` and `en` so it isn't a problem. Does yours not? If not, then that same logic needs to be applied to both places.

MaathavanJkr commented:
> OK, I fixed the conversion. I'll leave it to you to clean up the lint errors ;)
> 
> Now, what you need to do is say that if `aaa_bbb` is the only thing IE provides, we check `aaa_bbb` first and then `aaa`. My Chrome provides both `en-US` and `en` so it isn't a problem. Does yours not? If not, then that same logic needs to be applied to both places.
`aaa_bbb` is the only thing IE provides. my chrome also provides both

sushain97 commented:
> aaa_bbb is the only thing IE provides.

Cool. Then, can you modify the IE code to try `aaa_bbb` first and then if `isVariant(...)` is true, try `parentLang(...)` as well?

MaathavanJkr commented:
> > aaa_bbb is the only thing IE provides.
> 
> Cool. Then, can you modify the IE code to try `aaa_bbb` first and then if `isVariant(...)` is true, try `parentLang(...)` as well?

Already started to working on it. :)

sushain97 commented:
Awesome!

MaathavanJkr commented:
What about a language without a translation pair like Kyrgyz. You just removed `ValidSrcLang` function, I dont understand.

sushain97 commented:
Ah, right. My bad. I think you can revert that change!

This is why it would be nice to have tests...

sushain97 commented:
Please don't revert my changes. Your naming of these variables does not reflect their value. They don't contain BCP-47 codes, they contain ISO-639 codes.

MaathavanJkr commented:
Sorry, i just changed the `lang` variable name due to eslint error.

MaathavanJkr commented:
@sushain97  can u please change it to perfect variable name. :)

MaathavanJkr commented:
Is it Okay now?

sushain97 commented:
This looks fine! Sigh, I hate `var` scoping... 

Can you test it on IE/Chrome for the two things I mentioned?

MaathavanJkr commented:
Okay. :)

sushain97 commented on assets/js/translator.js:
```diff
@@ -1289,15 +1289,23 @@ function setDefaultSrcLang() {
         return iso639Lang;
     }
 
+    function setLangs(lang) {
```
This should already be done by the `restoreChoices('translator')`. This code shouldn't be messing with any of that logic.

MaathavanJkr replied:
This function happens after `restoreChoices('translator')`, so i thought we should check again

MaathavanJkr replied:
I think it wont mess with that as this function happens if `restoreChoices('translator')` doesnot happened.

sushain97 replied:
All this code happens before `restoreChoices('translator')`: https://github.com/apertium/apertium-html-tools/pull/352/files#diff-8331c23a07e5eb815c5b4393e81715efL467. This is really sketchy approach and will lead to hard to find bugs. Code outside that file shouldn't be touching local storage.

MaathavanJkr commented:
when im testing both instances, i got error when first visit as the `curSrcLang` is set but the UI not changed. (Fixed) and in IE the ISO639 isnt used. So I changed it to. And both instances are now working perfectly. :)

MaathavanJkr commented:
> There's a reason that `localization.js` doesn't export all those functions! They shouldn't be accessed here.

So what can we do? the UI isnt initialized when we set only curSrcLang.

sushain97 commented:
I'm fine with

```
curSrcLang = iso639CodesInverse[lang];
autoSelectDstLang();
```

Just not checking whether things exist in local storage first. If things do exist there, we should let `restoreChoices('translator')` which runs after us restore them and overwrite our changes. Does that not work?

MaathavanJkr commented:
> I'm fine with
> 
> ```
> curSrcLang = iso639CodesInverse[lang];
> autoSelectDstLang();
> ```
> 
> Just not checking whether things exist in local storage first. If things do exist there, we should let `restoreChoices('translator')` which runs after us restore them and overwrite our changes. Does that not work?

I think it will work. Lets try. :)

MaathavanJkr commented:
@sushain97 its working on both Chrome and IE. Can u please take a look at the code. :)

sushain97 commented:
I'm not exactly sure we're on the same page. I suggested leaving this as is with your new code setting `curSrcLang`, etc. Then, `restoreChoices('translator');` would run after `setDefaultSrcLang()` (as it already does), overwriting your settings.

MaathavanJkr commented:
> I'm not exactly sure we're on the same page. I suggested leaving this as is with your new code setting `curSrcLang`, etc. Then, `restoreChoices('translator');` would run after `setDefaultSrcLang()` (as it already does), overwriting your settings.

But it didnt work when running it after this function

sushain97 commented:
Why?

MaathavanJkr commented:
Im not sure with it, wait i will have look.

MaathavanJkr commented:
Got it! Because `restoreChoices('translator');` function doesnot set `curSrcLang` but changes the UI.

sushain97 commented:
https://github.com/apertium/apertium-html-tools/blob/0fe0c816c3527d33cab78681091d88c645a90f37/assets/js/persistence.js#L158 doesn't this do it?

MaathavanJkr commented:
> https://github.com/apertium/apertium-html-tools/blob/0fe0c816c3527d33cab78681091d88c645a90f37/assets/js/persistence.js#L158
> 
> doesn't this do it?

yeah, You are right.

I ran some tests and It seems the language we set to `curSrcLang` is set to Stored Lang when running function `autoSelectDstLang();`. then when we run `restoreChoices('translator')` its sets the same language. But i cant find where do that happens. Im trying...

MaathavanJkr commented:
Gotcha! In function `autoSelectDstLang();`, the function `handleNewCurrentLang()` is called. there the function `persistChoices()` is called which updates both `curSrcLang` and `curDstLang` to the local storage. 
If its right we can simple solve it by calling `handleNewCurrentLang()` after `restoreChoices('translator')`.

sushain97 commented:
Ah, jeez. Good detective work. I made some tweaks that I think should fix it?

Also, I wonder if we should be calling `handleNewCurrentLang`...

MaathavanJkr commented:
> Ah, jeez. Good detective work. I made some tweaks that I think should fix it?
> 
> Also, I wonder if we should be calling `handleNewCurrentLang`...

Yeah!! I tried this to do inside set `setDefaultLang();`, But this one is better. Thanks! :)

MaathavanJkr commented:
> Ah, jeez. Good detective work. I made some tweaks that I think should fix it?
> 
> Also, I wonder if we should be calling `handleNewCurrentLang` directly...

And i think we should call `handleNewCurrentLang` for curSrcLang under `setDefaultLang()`. as when im testing the UI isnt set when we ran for first visit.

sushain97 commented:
> And i think we should call handleNewCurrentLang for curSrcLang under setDefaultLang(). as when im testing, the UI isnt set when we ran for first visit.

Yep, that sounds like the right move after taking a quick look at the code. Could you make that tweak and test it?

MaathavanJkr commented:
> Yep, that sounds like the right move after taking a quick look at the code. Could you make that tweak and test it?

Done. :), Tested on both Chrome and IE for both instances and its working perfectly.

sushain97 commented:
Seems good!

sushain97 commented:
@jonorthwash This looks good on my end. Wanna try it before I merge? I don't have any particularly unique browser settings.
